File: game-core/src/deck.rs
use crate::models::Card;
use rand::seq::SliceRandom;

pub struct Deck {
    cards: Vec<Card>,
}

impl Deck {
    pub fn new() -> Self {
        let cards: Vec<Card> = ['S', 'H', 'D', 'C']
            .iter()
            .flat_map(|&suit| (2..=14).map(move |rank| Card::new(suit, rank)))
            .collect();

        let mut deck = Self { cards };
        deck.shuffle();
        deck
    }

    pub fn shuffle(&mut self) {
        self.cards.shuffle(&mut rand::thread_rng());
    }

    pub fn deal(&mut self, num_players: usize) -> Vec<Vec<Card>> {
        let mut hands: Vec<Vec<Card>> = vec![Vec::new(); num_players];

        for (i, card) in self.cards.drain(..).enumerate() {
            hands[i % num_players].push(card);
        }

        hands.iter_mut().for_each(|hand| hand.sort());
        hands
    }
}

File: game-core-py/hearts_game_core/deck.py
from hearts_game_core.game_models import Card
import random
from typing import List


class Deck:
    def __init__(self):
        self.cards = []
        self.reset()

    def reset(self):
        self.cards = [
            Card(suit=suit, rank=rank)
            for suit in "SHDC"
            for rank in range(2, 15)
        ]
        random.shuffle(self.cards)

    def shift_left(self, num_cards: int):
        self.cards = self.cards[num_cards:] + self.cards[:num_cards]    
        

    def deal(self, num_hands: int, num_cards: int) -> List[List[Card]]:
        hands = [self.cards[s * num_cards:(s + 1) * num_cards] for s in range(0, num_hands)]
        return hands
